packages/@react-aria/utils/src/useLoadMore.ts

```diff
@@ -0,0 +1,78 @@
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import {RefObject, useCallback, useRef} from 'react';
+import {useEvent} from './useEvent';
+// eslint-disable-next-line rulesdir/useLayoutEffectRule
+import {useLayoutEffect} from './useLayoutEffect';
+
+export interface LoadMoreProps {
+  /** Whether data is currently being loaded. */
+  isLoading?: boolean,
+  /** Handler that is called when more items should be loaded, e.g. while scrolling near the bottom.  */
+  onLoadMore?: () => void,
+  /**
+   * The amount of offset from the bottom of your scrollable region that should trigger load more.
+   * Uses a percentage value relative to the scroll body's client height. Load more is then triggered
+   * when your current scroll position's distance from the bottom of the currently loaded list of items is less than
+   * or equal to the provided value. (e.g. 1 = 100% of the scroll region's height).
+   * @default 1
+   */
+  scrollOffset?: number
+}
+
+export function useLoadMore(props: LoadMoreProps, ref: RefObject<HTMLElement | null>) {
+  let {isLoading, onLoadMore, scrollOffset = 1} = props;
+
+  // Handle scrolling, and call onLoadMore when nearing the bottom.
+  let isLoadingRef = useRef(isLoading);
+  let prevProps = useRef(props);
+  let onScroll = useCallback(() => {
+    if (ref.current && !isLoadingRef.current && onLoadMore) {
+      let shouldLoadMore = ref.current.scrollHeight - ref.current.scrollTop - ref.current.clientHeight < ref.current.clientHeight * scrollOffset;
+
+      if (shouldLoadMore) {
+        isLoadingRef.current = true;
+        onLoadMore();
+      }
+    }
+  }, [onLoadMore, ref, scrollOffset]);
+
+  useLayoutEffect(() => {
+    // Only update isLoadingRef if props object actually changed,
+    // not if a local state change occurred.
+    if (props !== prevProps.current) {
+      isLoadingRef.current = isLoading;
+      prevProps.current = props;
+    }
+
+    // TODO: this actually calls loadmore twice in succession on intial load because after the first load
+    // the scrollable element hasn't yet recieved its new height with the newly loaded items... Because of RAC collection needing two renders?
+    // Using ref.current.clientElementCount doesn't work either because the scrollable body may be the Table resize container (which contains the table, header and body: 3 children always)
+    // or the Table (which contains the header and body so 2 children) or the virtualizer in RSP (just contains the rows so has a variable count)
```

Comment:
Not quite sure the best way to go about handling this since the direct children of the scrollable element may not be the actual rows...

Reply:
Isn't `isLoadingRef` supposed to indicate if there is already a `onLoadMore()` call, which should prevent the second call? Did the removal of `wasLoading` introduce this? You confirmed this wasn't pre-exixisting?

Reply:
The core reason behind this is the two renders required for RAC collections to process new items: https://github.com/adobe/react-spectrum/pull/6494#discussion_r1670904973

High level flow that happens from what I can tell:
Table renders -> first time load is triggered (at this time the scrollable body's scrollHeight === clientHeight aka it isn't scrollable because no content) -> load finishes and returns first page of items, isLoading changes from true to false -> useLoadMore useLayoutEffect runs and collection begins processing the new items. scrollHeight of scrollable body is still the same as clientHeight since the collection hasn't done its second render yet and thus hasn't finished processing the collection so it hasn't rendered the new rows yet -> 2nd loadmore is triggered because of that

Reply:
These seems fine unless you think there is a performance issue to solve.